theano/gpuarray/opt.py

```diff
@@ -845,10 +848,14 @@ def local_gpu_pdbbreakpoint_op(node):
 def local_gpua_lazy_ifelse(op, context_name, inputs):
     if op.gpu:
         return
+    if isinstance(inputs[0].type, GpuArrayType):
+        return
     c = inputs[0]
     inps = []
     for v in inputs[1:]:
-        if isinstance(v.type, (tensor.TensorType, GpuArrayType)):
+        if isinstance(v.type, GpuArrayType):
+            return
+        elif isinstance(v.type, tensor.TensorType):
```

Comment:
as_gpuarray_variable will not create a new node if it is already on the GPU.

But you are right, if one input is on the GPU, we already moved that node to the GPU. So we can stop there. So maybe add a comment like:

    this node is already on GPU, so don't change the graph